i2c/imx6ull/libimx6ull-i2c.c

```diff
@@ -0,0 +1,361 @@
+
+
+/* clang-format off */
+/* NOTE: i2c registers are 16-bit and every second offset is reserved for future use */
+enum { iadr = 0, ifdr = 2, i2cr = 4, i2sr = 6, i2dr = 8 };
+/* clang-format on */
+
+static const uint32_t i2c_addr[] = { 0x021a0000, 0x021a4000, 0x021a8000, 0x021f8000 };
+static const int clocks[] = { pctl_clk_i2c1_serial, pctl_clk_i2c2_serial, pctl_clk_i2c3_serial, pctl_clk_i2c4_serial };
+
+
+typedef struct {
+	int pctl;
+	char val;
+} i2c_pctl_t;
+
+static const i2c_pctl_t i2c_pctl_mux[4][2] = {
+	{ { pctl_mux_gpio1_02, 0 }, { pctl_mux_gpio1_03, 0 } },
+	{ { pctl_mux_gpio1_00, 0 }, { pctl_mux_gpio1_01, 0 } },
+	{ { pctl_mux_lcd_d1, 4 }, { pctl_mux_lcd_d0, 4 } },
+	{ { pctl_mux_lcd_d3, 4 }, { pctl_mux_lcd_d2, 4 } },
+};
+
+static const i2c_pctl_t i2c_pctl_isel[4][2] = {
+	{ { pctl_isel_i2c1_scl, 0 }, { pctl_isel_i2c1_sda, 1 } },
+	{ { pctl_isel_i2c2_scl, 1 }, { pctl_isel_i2c2_sda, 1 } },
+	{ { pctl_isel_i2c3_scl, 2 }, { pctl_isel_i2c3_sda, 2 } },
+	{ { pctl_isel_i2c4_scl, 2 }, { pctl_isel_i2c4_sda, 2 } },
+};
+
+
+static const unsigned i2c_int_no[] = { 32 + 36, 32 + 37, 32 + 38, 32 + 35 };
+
+
+static struct {
+	volatile uint16_t *base;
+	uint16_t dev_no;
+
+	handle_t cond;
+	handle_t inth;
+	handle_t lock;
+} i2c = { 0 };
+
+/* Not implemented:
+ *  - slave mode
+ *  - multi-master mode (arbitration lost status flag inspection)
+ *  - register readout could be written by generating RESTART event instead of STOP / START (beneficial in multi-master mode)
+ */
+
+
+static void platform_init(void)
+{
+	platformctl_t pctl;
+	unsigned int i;
+
+	/* enable clock */
+	pctl.action = pctl_set;
+	pctl.type = pctl_devclock;
+	pctl.devclock.dev = clocks[i2c.dev_no - 1];
+	pctl.devclock.state = 0b11;
+
+	platformctl(&pctl);
+
+	/* set MUXes */
+	pctl.type = pctl_iomux;
+	pctl.iomux.sion = 1; /* from RM: SION bit needs to be enabled for i2c pads */
+
+	for (i = 0; i < 2; ++i) {
+		pctl.iomux.mux = i2c_pctl_mux[i2c.dev_no - 1][i].pctl;
+		pctl.iomux.mode = i2c_pctl_mux[i2c.dev_no - 1][i].val;
+		platformctl(&pctl);
+	}
+
+	/* set IOsel */
+	pctl.type = pctl_ioisel;
+
+	for (i = 0; i < 2; ++i) {
+		pctl.ioisel.isel = i2c_pctl_isel[i2c.dev_no - 1][i].pctl;
+		pctl.ioisel.daisy = i2c_pctl_isel[i2c.dev_no - 1][i].val;
+		platformctl(&pctl);
+	}
+
+	/* PAD configuration taken from linux Device Tree, not needed for our current targets, leaving commented-out
+	 * as it might need to be enabled for targets with missing pullup resistors (and written in generic way) */
+#if 0
+	pctl.type = pctl_iopad;
+	pctl.iopad.pad = pctl_pad_gpio1_02;
+	pctl.iopad.hys = 1;
+	pctl.iopad.pus = 0b10; /* 100kOhm pull-up */
+	pctl.iopad.pue = 0;    /* pull/keeper enabled */
+	pctl.iopad.pke = 1;
+	pctl.iopad.ode = 1;      /* open drain enable */
+	pctl.iopad.speed = 0b10; /* speed - 100 MHz */
+	pctl.iopad.dse = 0x4;
+	pctl.iopad.sre = 0;
+
+	platformctl(&pctl);
+
+	pctl.iopad.pad = pctl_pad_gpio1_03;
+	platformctl(&pctl);
+#endif
+}
+
+
+static int is_bus_busy(void)
```

Comment:
Use camelCase

Reply:
Ok, I assumed it is a requirement for external functions only